src/gui/AstroCalcDialog.cpp

```diff
@@ -2159,6 +2174,356 @@ void AstroCalcDialog::saveTransits()
 	}
 }
 
+void AstroCalcDialog::setLunarEclipseHeaderNames()
+{
+	lunareclipseHeader.clear();
+	lunareclipseHeader << q_("Date and Time");
+	lunareclipseHeader << q_("Saros number");
+	lunareclipseHeader << q_("Type");
+	lunareclipseHeader << q_("Gamma");
+	lunareclipseHeader << q_("Penumbral eclipse magnitude");
+	lunareclipseHeader << q_("Umbral eclipse magnitude");
+	ui->lunareclipseTreeWidget->setHeaderLabels(lunareclipseHeader);
+
+	// adjust the column width
+	for (int i = 0; i < LunarEclipseCount; ++i)
+	{
+		ui->lunareclipseTreeWidget->resizeColumnToContents(i);
+	}
+}
+
+void AstroCalcDialog::initListLunarEclipse()
+{
+	ui->lunareclipseTreeWidget->clear();
+	ui->lunareclipseTreeWidget->setColumnCount(LunarEclipseCount);
+	setLunarEclipseHeaderNames();
+	ui->lunareclipseTreeWidget->header()->setSectionsMovable(false);
+	ui->lunareclipseTreeWidget->header()->setDefaultAlignment(Qt::AlignHCenter);
+}
+
+QPair<double,double> AstroCalcDialog::getLunarEclipseXY() const
+{
+	// Algorithm adaped from Planet::getLunarEclipseMagnitudes() -- we need only x and y here.
+	// Find x, y of Besselian elements
+	QPair<double,double> LunarEclipseXY;
+	// Use geocentric coordinates
+	StelCore* core = StelApp::getInstance().getCore();
+	static SolarSystem* ssystem = GETSTELMODULE(SolarSystem);
+	core->setUseTopocentricCoordinates(false);
+	core->update(0);
+
+	double raMoon, deMoon, raSun, deSun;
+	StelUtils::rectToSphe(&raSun, &deSun, ssystem->getSun()->getEquinoxEquatorialPos(core));
+	StelUtils::rectToSphe(&raMoon, &deMoon, ssystem->getMoon()->getEquinoxEquatorialPos(core));
+
+	// R.A./Dec of Earth's shadow
+	const double raShadow = StelUtils::fmodpos(raSun + M_PI, 2.*M_PI);
+	const double deShadow = -(deSun);
+	const double raDiff = StelUtils::fmodpos(raMoon - raShadow, 2.*M_PI);
+
+	double x = cos(deMoon) * sin(raDiff);
+	x *= 3600. * M_180_PI;
+	double y = cos(deShadow) * sin(deMoon) - sin(deShadow) * cos(deMoon) * cos(raDiff);
+	y *= 3600. * M_180_PI;
+
+	LunarEclipseXY.first = x;
+	LunarEclipseXY.second = y;
+
+	return LunarEclipseXY;
+}
+
+void AstroCalcDialog::generateLunarEclipses()
+{
+	const bool onEarth = core->getCurrentPlanet()==solarSystem->getEarth();
+	if (onEarth) // Not sure it's right thing to do but should be ok.
+	{
+		initListLunarEclipse();
+
+		const double currentJD = core->getJD();   // save current JD
+		double startJD = StelUtils::qDateTimeToJd(QDateTime(ui->lunareclipseFromDateEdit->date()));
+		double stopJD = StelUtils::qDateTimeToJd(QDateTime(ui->lunareclipseToDateEdit->date()));
+		startJD = startJD - core->getUTCOffset(startJD) / 24.;
+		stopJD = stopJD - core->getUTCOffset(stopJD) / 24.;
+		int elements = static_cast<int>((stopJD - startJD) / 29.530588853);
```

Comment:
I think values `29.530588853` and `2451550.09765` should be defined as `const double` variables and just called in right places